nexus/src/context.rs

```diff
@@ -134,14 +134,24 @@ impl ServerContext {
         // nexus in dev for everyone
 
         // Set up DNS Client
-        let az_subnet =
-            Ipv6Subnet::<AZ_PREFIX>::new(config.deployment.subnet.net().ip());
-        info!(log, "Setting up resolver on subnet: {:?}", az_subnet);
-        let resolver = internal_dns::resolver::Resolver::new_from_subnet(
-            log.new(o!("component" => "DnsResolver")),
-            az_subnet,
-        )
-        .map_err(|e| format!("Failed to create DNS resolver: {}", e))?;
+        let resolver = match config.deployment.internal_dns {
+            nexus_config::InternalDns::FromSubnet { subnet } => {
+                let az_subnet = Ipv6Subnet::<AZ_PREFIX>::new(subnet.net().ip());
+                info!(log, "Setting up resolver on subnet: {:?}", az_subnet);
+                internal_dns::resolver::Resolver::new_from_subnet(
+                    log.new(o!("component" => "DnsResolver")),
+                    az_subnet,
+                )
+                .map_err(|e| format!("Failed to create DNS resolver: {}", e))?
+            }
+            nexus_config::InternalDns::FromAddress { address } => {
+                internal_dns::resolver::Resolver::new_from_addrs(
+                    log.new(o!("component" => "DnsResolver")),
+                    vec![address],
+                )
+                .map_err(|e| format!("Failed to create DNS resolver: {}", e))?
+            }
```

Comment:
This is the chunk that's new - tests can use internal DNS without squatting on the whole subnet.

Reply:
Before this PR, we passed an argument as a part of Nexus' `DeploymentConfig`:

https://github.com/oxidecomputer/omicron/blob/9b85b6190cf58e7860a04471f506fbaecabd219f/common/src/nexus_config.rs#L118-L119

This "subnet" argument was then used to determine the address of DNS server(s):

https://github.com/oxidecomputer/omicron/blob/9b85b6190cf58e7860a04471f506fbaecabd219f/nexus/src/context.rs#L136-L144

This is how I expect we'll at least bootstrap the discovery of internal DNS servers on the underlay.

However, this means that:
- Given an input subnet...
- ... Nexus assumed it could infer the DNS server(s) from that subnet, by knowing that they'd reliably be at particular reserved addresses / ports.
- ... But in tests, this means we'd need full control over everything running in a subnet. This breaks down if a test wants to try placing a internal DNS server on "localhost" (for example, letting the host OS decide the port) instead of a reserved address within a subnet.